packages/api/amplify_api/lib/src/graphql/ws/web_socket_connection.dart

```diff
@@ -0,0 +1,240 @@
+  var _initMemo = AsyncMemoizer<void>();
+  Completer<void> _connectionReady = Completer<void>();
+
+  /// Fires when the connection is ready to be listened to after the first
+  /// `connection_ack` message.
+  Future<void> get ready => _connectionReady.future;
+
+  /// {@macro amplify_api.ws.web_socket_connection}
+  WebSocketConnection(this._config, this._authProviderRepo,
+      {required AmplifyLogger logger})
+      : _logger = logger;
+
+  /// Connects _subscription stream to _onData handler.
+  @visibleForTesting
+  StreamSubscription<WebSocketMessage> getStreamSubscription(
+      Stream<dynamic> stream) {
+    return stream
+        .transform(const WebSocketMessageStreamTransformer())
+        .listen(_onData, onError: (Object e) {
+      _connectionError(ApiException('Connection failed.',
+          underlyingException: e.toString()));
+    });
+  }
+
+  /// Connects WebSocket channel to _subscription stream but does not send connection
+  /// init message.
+  @visibleForTesting
+  Future<void> connect(Uri connectionUri) async {
+    _channel = WebSocketChannel.connect(
+      connectionUri,
+      protocols: webSocketProtocols,
+    );
+    _subscription = getStreamSubscription(_channel!.stream);
+  }
+
+  void _connectionError(ApiException exception) {
+    _connectionReady.completeError(_connectionError);
+    _channel?.sink.close();
+    _resetConnectionInit();
+  }
+
+  // Reset connection init variables so it can be re-attempted.
+  void _resetConnectionInit() {
+    _initMemo = AsyncMemoizer<void>();
+    _connectionReady = Completer<void>();
+  }
+
+  /// Closes the WebSocket connection and cleans up local variables.
+  @override
+  void close([int closeStatus = _defaultCloseStatus]) {
+    _logger.verbose('Closing web socket connection.');
+    final reason =
+        closeStatus == _defaultCloseStatus ? 'client closed' : 'unknown';
+    _subscription?.cancel();
+    _channel?.sink.done.whenComplete(() => _channel = null);
+    _channel?.sink.close(closeStatus, reason);
+    _rebroadcastController.close();
+    _timeoutTimer?.cancel();
+    _resetConnectionInit();
+  }
+
+  /// Initializes the connection.
+  ///
+  /// Connects to WebSocket, sends connection message and resolves future once
+  /// connection_ack message received from server. If the connection was previously
+  /// established then will return previously completed future.
+  Future<void> init() => _initMemo.runOnce(_init);
+
+  Future<void> _init() async {
+    final connectionUri =
+        await generateConnectionUri(_config, _authProviderRepo);
+    await connect(connectionUri);
+
+    send(WebSocketConnectionInitMessage());
+
+    return ready;
+  }
+
+  /// Subscribes to the given GraphQL request. Returns the subscription object,
+  /// or throws an [Exception] if there's an error.
+  Stream<GraphQLResponse<T>> subscribe<T>(
+    GraphQLRequest<T> request,
+    void Function()? onEstablished,
+  ) {
+    final subscriptionId = uuid();
```

Comment:
nit: This might as well be changed to use request.id now. It'll be that way in reconnect PR

Reply:
yes good catch